src/Shared/FileUtilities.cs

```diff
@@ -1054,23 +1055,89 @@ internal static string WithTrailingSlash(this string s)
 
         internal static string NormalizeForPathComparison(this string s) => s.ToSlash().TrimTrailingSlashes();
 
+        // TODO: assumption on file system case sensitivity: https://github.com/Microsoft/msbuild/issues/781
         internal static bool PathsEqual(string path1, string path2)
         {
-            var trim1 = path1.TrimTrailingSlashes();
-            var trim2 = path2.TrimTrailingSlashes();
-
-            if (string.Equals(trim1, trim2, PathComparison))
+            if (path1 == null && path2 == null)
             {
                 return true;
             }
+            if (path1 == null || path2 == null)
+            {
+                return false;
+            }
+
+            var endA = path1.Length - 1;
+            var endB = path2.Length - 1;
+
+            // Trim trailing slashes
+            for (var i = endA; i >= 0; i--)
+            {
+                var c = path1[i];
+                if (c == '/' || c == '\\')
+                {
+                    endA--;
+                }
+                else
+                {
+                    break;
+                }
+            }
 
-            var slash1 = trim1.ToSlash();
-            var slash2 = trim2.ToSlash();
+            for (var i = endB; i >= 0; i--)
+            {
+                var c = path2[i];
+                if (c == '/' || c == '\\')
+                {
+                    endB--;
+                }
+                else
+                {
+                    break;
+                }
+            }
 
-            return string.Equals(slash1, slash2, PathComparison);
+            if (endA != endB)
+            {
+                // Lengths not the same
+                return false;
+            }
+
+            for (var i = 0; i <= endA; i++)
+            {
+                var charA = (uint)path1[i];
+                var charB = (uint)path2[i];
+
+                if ((charA | charB) > 0x7F)
+                {
+                    // Non-ascii chars move to non fast path
+                    return PathsEqualNonAscii(path1, path2, i, endA - i + 1);
```

Comment:
This might be a problem for other languages, where there is a high chance to get unicode characters in paths (e.g. user names), which would make globbing deep directories faster for English users, but keep it slow for non-English ones (#2238 suggests PathsEqual might be a big perf hit when globbing over node_modules).

Is there any issue with doing the expensive check just for the characters in the current iteration (using Char.ToUpper to get culture aware case insensitivity) rather than falling back to the expensive check for the rest of the strings?